### PR TITLE
[Validation] Removed validation of `DisableSimultaneousMultithreading` and adapted disabling logic accordingly.

### PR DESCRIPTION
**Pull request by gmarciani**
### Description of changes
1. Removed validation of `DisableSimultaneousMultithreading` in order to allow the disablement of multi-threading on all instances supporting it or to skip it on instances not supporting it.
2. Adapted internal logic to select disablement via CpuOptions or manually

Proved via scripting that single-threading can be set via `CpuOptions` for all instances having 1 within `VcpuInfo.ValidThreadsPerCore`, even if they have `VcpuInfo.DefaultThreadsPerCore=1` and regardless the supported architecture is `x86_64` or `arm64`.

1. DefaultThreadsPerCore > 1 AND ValidThreadsPerCore contains 1 => multi-threading must be disabled via CpuOptions
1. DefaultThreadsPerCore > 1 AND ValidThreadsPerCore not reported => multi-threading must be disabled manually
1. DefaultThreadsPerCore = 1 AND ValidThreadsPerCore contains 1 => nothing to do
1. DefaultThreadsPerCore = 1 AND ValidThreadsPerCore not reported => nothing to do

### Tests
1. Unit tests succeeded
3. Verified that the criteria to select the instances supporting the single threading via CpuOptions is correct by setting it for every instance types in DUB.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

Signed-off-by: Giacomo Marciani <mgiacomo@amazon.com>

### CODE REVIEW

File: tests/integration-tests/tests/efa/test_efa.py [hunk 1]
```diff
@@ -54,7 +54,7 @@ def test_efa(
         multithreading_disabled = True
     else:
         head_node_instance = "c6g.16xlarge"
-        multithreading_disabled = False
+        multithreading_disabled = True
```
Comment (lukeseawalker) [inline tests/integration-tests/tests/efa/test_efa.py]: what about removing this variable if there is only one case? Than you can also change the config to remove the placeholder

File: cli/tests/pcluster/aws/dummy_aws_api.py [hunk 2]
```diff
@@ -57,7 +55,14 @@ def max_network_interface_count(self):
...
 
     def default_threads_per_core(self):
-        return self._default_threads_per_core
+        # There are more instance types, but for the simplicity of the mock,
+        # we consider only t2 as having one thread per core.
+        return 1 if self._instance_type.startswith("t2") else 2
+
+    def valid_threads_per_core(self):
+        # There are more instance types, but for the simplicity of the mock,
+        # we consider only t2 as having one thread per core.
```
Comment (lukeseawalker) [inline cli/tests/pcluster/aws/dummy_aws_api.py]: "we consider only t2 as having one thread per core."
probably you meant to consider only t2 has not reporting valid thread per core

File: cli/tests/pcluster/validators/test_all_validators/test_validators_are_called_with_correct_argument/slurm.yaml [hunk 3]
```diff
@@ -30,8 +30,10 @@ Scheduling:
         - Name: compute_resource1
           InstanceType: c5.4xlarge
           MaxCount: 5
+          DisableSimultaneousMultithreading: true
```
Comment (lukeseawalker) [inline cli/tests/pcluster/validators/test_all_validators/test_validators_are_called_with_correct_argument/slurm.yaml:33]: is it worth adding the case "instance not supporting multi-thread disablement" + DisableSimultaneousMultithreading = true?, i.e.
```
InstanceType: t2.large
DisableSimultaneousMultithreading: true
```

File: CHANGELOG.md [hunk 4]
```diff
@@ -39,6 +39,8 @@ CHANGELOG
 **BUG FIXES**
 - Fix default for disable validate and test components when building custom AMI. The default was to disable those components, but it wasn't effective.
 - Handle corner case in the scaling logic when instance is just launched and the describe instances API doesn't report yet all the EC2 info.
+- Fix validation for configuration parameter `Scheduling/SlurmQueues/ComputeResources/DisableSimultaneousMultithreading`
```
Comment (lukeseawalker) [inline CHANGELOG.md]: isn't the DisableSimultaneousMultithreading fixed also for the HeadNode?

File: tests/integration-tests/tests/efa/test_efa.py (hunk 1 above)
Comment (gmarciani) [reply tests/integration-tests/tests/efa/test_efa.py]: Fixed in the next rev.

File: cli/tests/pcluster/aws/dummy_aws_api.py (hunk 2 above)
Comment (gmarciani) [reply cli/tests/pcluster/aws/dummy_aws_api.py]: Yes, copy-paste error. Rephrasing in the next rev.

File: cli/tests/pcluster/validators/test_all_validators/test_validators_are_called_with_correct_argument/slurm.yaml (hunk 3 above)
Comment (gmarciani) [reply cli/tests/pcluster/validators/test_all_validators/test_validators_are_called_with_correct_argument/slurm.yaml:33]: Good point, thanks. Added in the next rev.

File: CHANGELOG.md (hunk 4 above)
Comment (gmarciani) [reply CHANGELOG.md]: Good point, thanks! It is. I would put a general 

"Fix validation for configuration parameter `DisableSimultaneousMultithreading` to allow the disablement of multi-threading on all instances supporting it." then

File: cli/src/pcluster/aws/aws_resources.py [hunk 5]
```diff
@@ -224,10 +224,13 @@ def instance_type(self):
...
 
     def is_cpu_options_supported_in_lt(self):
-        """Check whether hyperthreading can be disabled via CPU options."""
-        # If default threads per core is 1, HT doesn't need to be disabled
+        """Check whether CPUOptions can be set in launch template."""
         # When CpuOptions is not supported, valid threads per core can't be found in vcpu info
-        return self.default_threads_per_core() > 1 and 1 in self.valid_threads_per_core()
+        return len(self.valid_threads_per_core()) != 0
+
+    def is_multi_threading_disablement_supported(self):
```
Comment (lukeseawalker) [inline cli/src/pcluster/aws/aws_resources.py]: I think that the logic on the method "disable_simultaneous_multithreading_manually" should be revisited to account this change. Can you please check and in case add also a test to cover this case? I'm thinking about for example a metal instance with DefaultThreadsPerCore = 2 but that doesn't support cpu_options.

File: cli/src/pcluster/aws/aws_resources.py [hunk 6]
```diff
@@ -224,10 +224,13 @@ def instance_type(self):
         return self.instance_type_data.get("InstanceType")
 
     def is_cpu_options_supported_in_lt(self):
```
Comment (lukeseawalker) [inline cli/src/pcluster/aws/aws_resources.py:226]: since you have changed the method behavior, please also check all the callers of this method, to understand if they need to be changed as well.

Comment (gmarciani) [reply cli/src/pcluster/aws/aws_resources.py:226]: Thanks, done in the next rev.

File: cli/src/pcluster/aws/aws_resources.py (hunk 5 above)
Comment (gmarciani) [reply cli/src/pcluster/aws/aws_resources.py]: We decided to remove the validator because we have the following cases:
1. DefaultThreadsPerCore > 1 AND ValidThreadsPerCore contains 1 => multi-threading must be disabled via CpuOptions
1. DefaultThreadsPerCore > 1 AND ValidThreadsPerCore not reported => multi-threading must be disabled manually
1. DefaultThreadsPerCore = 1 AND ValidThreadsPerCore contains 1 => nothing to do
1. DefaultThreadsPerCore = 1 AND ValidThreadsPerCore not reported => nothing to do

So there is no case where the validators should fail.

File: CHANGELOG.md [hunk 7]
```diff
@@ -39,6 +39,8 @@ CHANGELOG
 **BUG FIXES**
 - Fix default for disable validate and test components when building custom AMI. The default was to disable those components, but it wasn't effective.
 - Handle corner case in the scaling logic when instance is just launched and the describe instances API doesn't report yet all the EC2 info.
+- Removed validation of `DisableSimultaneousMultithreading` in order to allow the disablement of multi-threading on all instances supporting it
+  or to skip it on instances not supporting it.
```
Comment (lukeseawalker) [inline CHANGELOG.md]: " skip it on instances not supporting it." we are disabling multithreading also in the case of instance not supporting it, through the manual disable flag.
I would rewrite the bullet with something "Dropped validation that would prevent ARM instance type to be used when DisableSimultaneousMultithreading was set to true."

File: cli/src/pcluster/config/cluster_config.py [hunk 8]
```diff
@@ -1022,12 +1016,20 @@ def instance_type_info(self) -> InstanceTypeInfo:
     @property
     def disable_simultaneous_multithreading_via_cpu_options(self) -> bool:
         """Return true if simultaneous multithreading must be disabled through cpu options."""
-        return self.disable_simultaneous_multithreading and self.instance_type_info.is_cpu_options_supported_in_lt()
+        return (
+            self.disable_simultaneous_multithreading
+            and self.instance_type_info.default_threads_per_core() > 1
```
Comment (lukeseawalker) [inline cli/src/pcluster/config/cluster_config.py:1021]: What about removing "and self.instance_type_info.default_threads_per_core() > 1"?
This will semplify the logic and will use cpu options even in the case default_threads_per_core = 1

File: CHANGELOG.md (hunk 7 above)
Comment (gmarciani) [reply CHANGELOG.md]: Totally agree with you. Your proposal is a lot more customer-centric, thanks

File: cli/src/pcluster/config/cluster_config.py (hunk 8 above)
Comment (gmarciani) [reply cli/src/pcluster/config/cluster_config.py:1021]: If this is not a strong concern for you, I think we should keep this conditions as it is because it guarantees that we don't do any unnecessary setup action, whether it is manual or via template.

Furthermore, since sooner or later we are moving towards a scenario where we will perform this specific disablement only manually, it would be even better to skip manual steps when not required.

What do you think?

Comment (lukeseawalker) [reply cli/src/pcluster/config/cluster_config.py:1021]: ACK

**Review by lukeseawalker — APPROVED** [review]
